File: ios/plaid_flutter/Source/plaid_flutter/Classes/TopViewControllerHelper.m
#import "TopViewControllerHelper.h"

@implementation UIApplication (TopViewController)

- (UIViewController *)topViewControllerForWindow:(UIWindow *)window {
    UIViewController *topViewController = window.rootViewController;
    while (topViewController.presentedViewController) {
        topViewController = topViewController.presentedViewController;
    }
    
    if ([topViewController isKindOfClass:[UINavigationController class]]) {
        topViewController = [(UINavigationController *)topViewController topViewController];
    } else if ([topViewController isKindOfClass:[UITabBarController class]]) {
        topViewController = [(UITabBarController *)topViewController selectedViewController];
    }
    
    return topViewController;
}

- (UIViewController *)topViewController {
    // Handle multi-window/scene
    UIScene *scene = [UIApplication sharedApplication].connectedScenes.anyObject;
    if ([scene isKindOfClass:[UIWindowScene class]]) {
        UIWindowScene *windowScene = (UIWindowScene *)scene;
        for (UIWindow *window in windowScene.windows) {
            if (window.isKeyWindow) {
                return [self topViewControllerForWindow:window];
            }
        }
    }
    return nil;
}

@end

File: ios/plaid_flutter/Sources/plaid_flutter/TopViewControllerHelper.swift
import UIKit

extension UIApplication {
  var plaidTopViewController: UIViewController? {
    let foregroundScenes =
      connectedScenes
      .compactMap { $0 as? UIWindowScene }
      .filter { $0.activationState == .foregroundActive }

    let window =
      foregroundScenes
      .compactMap { scene in
        scene.windows.first(where: { $0.isKeyWindow })
          ?? scene.windows.first(where: { !$0.isHidden })
      }
      .first

    return window?.rootViewController?.plaidTopViewController
  }
}

extension UIViewController {
  var plaidTopViewController: UIViewController {
    if let presentedViewController,
      !presentedViewController.isBeingDismissed
    {
      return presentedViewController.plaidTopViewController
    }

    if let navigationController = self as? UINavigationController,
      let visibleViewController = navigationController.visibleViewController
    {
      return visibleViewController.plaidTopViewController
    }

    if let tabBarController = self as? UITabBarController,
      let selectedViewController = tabBarController.selectedViewController
    {
      return selectedViewController.plaidTopViewController
    }

    if let splitViewController = self as? UISplitViewController,
      let lastViewController = splitViewController.viewControllers.last
    {
      return lastViewController.plaidTopViewController
    }

    return self
  }
}

extension UIView {
  var plaidOwningViewController: UIViewController? {
    var responder: UIResponder? = self
    while let nextResponder = responder?.next {
      if let viewController = nextResponder as? UIViewController {
        return viewController
      }
      responder = nextResponder
    }
    return nil
  }
}
